File: 2020/day02/day02.py
#!/usr/bin/env python3
import sys
import re
from collections import namedtuple

if len(sys.argv) > 1:
    filename = sys.argv[1]
else:
    filename = "input.txt"

Password = namedtuple("Password", ["min", "max", "req", "password"])

passwords = []
for i in open(filename).read().splitlines():
    match = re.match(r"^(\d+)-(\d+) (.): (.+)$", i)
    min_count, max_count = int(match.group(1)), int(match.group(2))
    passwords.append(
        Password(min_count, max_count, match.group(3), match.group(4)))


def is_valid_star1(password):
    count = password.password.count(password.req)
    return password.min <= count <= password.max


def is_valid_star2(password):
    first = password.password[password.min - 1] == password.req
    second = password.password[password.max - 1] == password.req
    return first ^ second


# star 1
count = 0
for password in passwords:
    if is_valid_star1(password):
        count += 1
print(count)

# star 2
count = 0
for password in passwords:
    if is_valid_star2(password):
        count += 1
print(count)

File: 2020/day02/day02.hs
#!/usr/bin/env stack
-- stack --resolver lts-14.16 script

import System.Environment (getArgs)
import Text.Parsec
import Text.Parsec.Char
import Text.Parsec.String

data Password = Password
  { first :: Int
  , second :: Int
  , req :: Char
  , password :: String
  } deriving (Show)

parsePassword :: Parser Password
parsePassword = do
  f <- read <$> many digit
  char '-'
  s <- read <$> many digit
  spaces
  r <- anyChar
  char ':'
  spaces
  p <- many $ noneOf "\n"
  return $ Password f s r p

parsePasswords :: Parser [Password]
parsePasswords = many $ parsePassword <* optional newline

star1 :: [Password] -> Int
star1 = length . filter isValid
  where count y = foldl (\acc x -> if x == y then acc + 1 else acc) 0
        isValid p =
          let c = count (req p) (password p)
          in first p <= c && c <= second p

star2 :: [Password] -> Int
star2 = length . filter isValid
  where isValid p =
          let f = password p !! (first p - 1) == req p
              s = password p !! (second p - 1) == req p
          in f && not s || not f && s

main :: IO ()
main = do
  args <- getArgs
  let
    filename = case args of
      (x:_) -> x
      _     -> "input.txt"
  (Right input) <- parse parsePasswords "" <$> readFile filename
  print $ star1 input
  print $ star2 input
